tests/saml/test_controller.py

```diff
@@ -0,0 +1,123 @@
+from api.saml.metadata import ServiceProviderMetadata, UIInfo, NameIDFormat, Service, IdentityProviderMetadata
+from api.saml.provider import SAMLWebSSOAuthenticationProvider
+from tests.saml import fixtures
+from tests.saml.controller_test import ControllerTest
+from tests.saml.test_auth import SAML_RESPONSE
+
+SERVICE_PROVIDER = ServiceProviderMetadata(
+    fixtures.SP_ENTITY_ID,
+    UIInfo(),
+    NameIDFormat.UNSPECIFIED.value,
+    Service(fixtures.SP_ACS_URL, fixtures.SP_ACS_BINDING)
+)
+
+IDENTITY_PROVIDERS = [
+    IdentityProviderMetadata(
+        fixtures.IDP_1_ENTITY_ID,
+        UIInfo(),
+        NameIDFormat.UNSPECIFIED.value,
+        Service(fixtures.IDP_1_SSO_URL, fixtures.IDP_1_SSO_BINDING),
+        signing_certificates=[
+            fixtures.SIGNING_CERTIFICATE
+        ]
+    ),
+    IdentityProviderMetadata(
+        fixtures.IDP_2_ENTITY_ID,
+        UIInfo(),
+        NameIDFormat.UNSPECIFIED.value,
+        Service(fixtures.IDP_2_SSO_URL, fixtures.IDP_2_SSO_BINDING)
+    )
+]
+
+
+class SAMLControllerTest(ControllerTest):
+    def test_saml_authentication_redirect(self):
+        configuration = create_autospec(spec=SAMLConfiguration)
+        configuration.get_debug = MagicMock(return_value=False)
+        configuration.get_strict = MagicMock(return_value=False)
+        configuration.get_service_provider = MagicMock(return_value=SERVICE_PROVIDER)
+        configuration.get_identity_providers = MagicMock(return_value=IDENTITY_PROVIDERS)
+        onelogin_configuration = SAMLOneLoginConfiguration(configuration)
+        authentication_manager = SAMLAuthenticationManager(onelogin_configuration)
+
+        authentication_manager_factory = create_autospec(spec=SAMLAuthenticationManagerFactory)
+        authentication_manager_factory.create = MagicMock(return_value=authentication_manager)
+
+        with patch('api.saml.provider.SAMLAuthenticationManagerFactory', autospec=True) \
+                as authentication_manager_factory_constructor:
+            authentication_manager_factory_constructor.return_value = authentication_manager_factory
+
+            provider = SAMLWebSSOAuthenticationProvider(self._library, self._integration)
+            authenticator = Authenticator(self._db)
+
+            authenticator.library_authenticators['default'].register_saml_provider(provider)
+
+            controller = SAMLController(self.app.manager, authenticator)
+
+            query = urllib.urlencode({
+                'provider': SAMLWebSSOAuthenticationProvider.NAME,
+                'idp_entity_id': IDENTITY_PROVIDERS[0].entity_id
+            })
+
+            with self.app.test_request_context('/saml_authenticate?' + query):
+                request.library = self._default_library
+                result = controller.saml_authentication_redirect(request.args, self._db)
+
+                print(result)
+
+                eq_(result.status_code, 302)
+                location_header = result.headers.get('Location')
```

Comment:
Can you make any assertions about the value of the Location header here?

Reply:
Added additional tests for `SAMLController`